metals/src/main/scala/scala/meta/internal/metals/NewFilesProvider.scala

```diff
@@ -7,51 +7,120 @@ import MetalsEnrichments._
+    MetalsPickItem(id = "worksheet", label = "Worksheet")
+
+  def createNewFileDialog(directoryUri: Option[URI])(
       implicit ec: ExecutionContext
-  ): Future[URI] = kind match {
-    case "class" | "object" | "trait" =>
-      createClass(directory, name, kind)
-    case "package-object" =>
-      createPackageObject(directory)
-    case "worksheet" =>
-      createWorksheet(directory, name)
-    case invalid => Future.failed(new IllegalArgumentException(invalid))
+  ): Future[Unit] = {
+    val directory = directoryUri
+      .map(_.toString.toAbsolutePath)
+      .orElse(focusedDocument.map(_.parent))
+
+    val newlyCreatedFile =
+      askForKind
+        .continueWith {
+          case kind @ (classPick.id | objectPick.id | traitPick.id) =>
+            askForName(kind)
+              .endWith(
+                createClass(directory, _, kind)
+              )
+          case worksheetPick.id =>
+            askForName(worksheetPick.id)
+              .endWith(
+                createWorksheet(directory, _)
+              )
+          case packageObjectPick.id =>
+            endWith(
+              createPackageObject(directory)
+            )
+          case invalid =>
+            Future.failed(new IllegalArgumentException(invalid))
+        }
+
+    newlyCreatedFile.map {
+      case Some(path) =>
+        openFile(path)
+      case None => ()
+    }
   }
 
-  def createWorksheet(directory: Option[URI], name: String)(
+  private def askForKind(
       implicit ec: ExecutionContext
-  ): Future[URI] = {
-    val path = directory
-      .fold(workspace)(_.toString.toAbsolutePath)
-      .resolve(name + ".worksheet.sc")
-    createFile(path)
+  ): Future[Option[String]] = {
+    client
+      .metalsPickInput(
+        MetalsPickInputParams(
+          List(
+            classPick,
+            objectPick,
+            traitPick,
+            packageObjectPick,
+            worksheetPick
+          ).asJava,
+          placeHolder = NewScalaFile.selectTheKindOfFileMessage
+        )
+      )
+      .asScala
+      .map {
+        case kind if !kind.cancelled => Some(kind.itemId)
+        case _ => None
+      }
   }
 
-  def createClass(directory: Option[URI], name: String, kind: String)(
+  private def askForName(kind: String)(
       implicit ec: ExecutionContext
-  ): Future[URI] = {
-    val path = directory
-      .fold(workspace)(_.toString.toAbsolutePath)
-      .resolve(name + ".scala")
+  ): Future[Option[String]] = {
+    client
+      .metalsInputBox(
+        MetalsInputBoxParams(prompt = NewScalaFile.enterNameMessage(kind))
+      )
+      .asScala
+      .map {
+        case name if !name.cancelled => Some(name.value)
+        case _ => None
+      }
+  }
+
+  private def createClass(
+      directory: Option[AbsolutePath],
+      name: String,
+      kind: String
+  )(
+      implicit ec: ExecutionContext
+  ): Future[AbsolutePath] = {
+    val path = directory.getOrElse(workspace).resolve(name + ".scala")
```

Comment:
What happens if the name is `a/b/c/A` - I think we should create the directories there. Most likely `createFile` should also create directories? Package will then be added automatically which is neat.

Reply:
it fails. Not really sure, it looks neat to ask for the name for the new class `foo/Bar` to create `class Bar`at `foo/Bar.scala`

Reply:
Could we implement it here? I think we just need to do `getParent.mkdirs`